File: web/src/api/authoring.ts
import { useMutation, useQuery } from '@tanstack/react-query'
import { apiFetch } from './client'
import type { TestInput } from './tests'

/** AI test authoring + auto-discovery (S26). Everything PROPOSES — nothing is
 *  created until the user confirms via the test API. */
export interface TestSpec {
  name: string
  type: string
  target: string
  interval_seconds: number
  timeout_seconds: number
  params?: Record<string, string>
  enabled: boolean
}

export interface TestProposal {
  spec: TestSpec
  rationale?: string
  source: string
}

export interface DiscoverProposal {
  spec: TestSpec
  rationale: string
  score: number
  source: string
}

/** useAuthorTest turns a natural-language prompt into a schema-valid proposal. */
export function useAuthorTest() {
  return useMutation({
    mutationFn: (prompt: string) =>
      apiFetch<TestProposal>('/ai/author', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ prompt }),
      }),
  })
}

/** useDiscover proposes monitorable targets mined from observed telemetry. */
export function useDiscover() {
  return useQuery({
    queryKey: ['ai', 'discover'],
    queryFn: () =>
      apiFetch<{ proposals: DiscoverProposal[] }>('/ai/discover', { method: 'POST' }).then(
        (r) => r.proposals,
      ),
  })
}

/** specToInput maps a proposed spec onto the test-create input. */
export function specToInput(s: TestSpec): TestInput {
  return {
    name: s.name,
    type: s.type,
    target: s.target,
    interval_seconds: s.interval_seconds,
    timeout_seconds: s.timeout_seconds,
    enabled: true,
  }
}

File: internal/control/authoring.go
// SPDX-License-Identifier: LicenseRef-probectl-TBD

package control

import (
	"context"
	"errors"
	"log/slog"
	"net/http"
	"strings"

	"github.com/imfeelingtheagi/probectl/internal/ai"
	"github.com/imfeelingtheagi/probectl/internal/ai/author"
	"github.com/imfeelingtheagi/probectl/internal/apierror"
	"github.com/imfeelingtheagi/probectl/internal/auth"
	"github.com/imfeelingtheagi/probectl/internal/config"
	"github.com/imfeelingtheagi/probectl/internal/store"
	"github.com/imfeelingtheagi/probectl/internal/tenancy"
)

// buildAuthor wires the AI test-authoring engine (S26). It uses the configured
// model when one is set (reusing the S24 model config) and otherwise the
// deterministic, air-gapped heuristic — so authoring works with zero external
// calls by default.
func buildAuthor(cfg *config.Config, log *slog.Logger, gate *ai.EgressGate) *author.Engine {
	m := buildModel(cfg, log)
	if c, ok := m.(ai.RemoteCompleter); ok {
		// AIRCA-005: the authoring model rides the SAME egress gate as RCA
		// and MCP — per-tenant consent, redaction, audit. A remote authoring
		// call without consent is denied; the heuristic author still works.
		return author.NewEngine(author.NewModelAuthor(ai.NewGatedCompleter(c, gate), m.Name()))
	}
	return author.NewEngine(author.HeuristicAuthor{})
}

// --- /v1/ai/author + /v1/ai/discover handlers (propose only — never auto-apply) ---

type authorRequest struct {
	Prompt string `json:"prompt"`
}

// handleAIAuthor turns a natural-language request into a schema-valid test config
// pending the user's confirmation. It NEVER creates the test (CLAUDE.md §7
// guardrail 8 — propose, human-gated); the user applies it via POST /v1/tests.
func (s *Server) handleAIAuthor(w http.ResponseWriter, r *http.Request) error {
	var req authorRequest
	if err := decodeJSON(r, &req); err != nil {
		return err
	}
	prompt := strings.TrimSpace(req.Prompt)
	if prompt == "" || len(prompt) > 2000 {
		return apierror.Validation("prompt is required (1–2000 characters)")
	}
	p := auth.PrincipalFrom(r.Context())
	if p == nil {
		return apierror.Unauthorized("authentication required")
	}
	proposal, err := s.authorEngine.Author(r.Context(), prompt)
	if err != nil {
		if errors.Is(err, author.ErrModelUnavailable) {
			s.log.Warn("ai author model unavailable", "error", err)
			return apierror.Unavailable("the authoring model is temporarily unavailable")
		}
		// ErrCannotAuthor (or an invalid generated config) → a 422 with guidance.
		return apierror.Validation(err.Error())
	}
	if s.pool != nil {
		if auditErr := s.inTenant(r, func(ctx context.Context, sc tenancy.Scope) error {
			return s.recordAudit(ctx, sc, r, "ai.author", "", map[string]any{"type": proposal.Spec.Type, "target": proposal.Spec.Target})
		}); auditErr != nil {
			s.log.Warn("audit ai.author failed", "error", auditErr)
		}
	}
	writeJSON(w, http.StatusOK, proposal)
	return nil
}

// handleAIDiscover proposes monitorable targets mined from the tenant's observed
// telemetry (incident targets today; the eBPF service map / flows / BGP / DNS plug
// into the same Observation input as those sources are wired), ranked, thresholded
// to avoid noise, and deduped against existing tests. Proposals only.
func (s *Server) handleAIDiscover(w http.ResponseWriter, r *http.Request) error {
	if auth.PrincipalFrom(r.Context()) == nil {
		return apierror.Unauthorized("authentication required")
	}
	var obs []author.Observation
	var existing []string
	if s.pool != nil {
		if err := s.inTenant(r, func(ctx context.Context, sc tenancy.Scope) error {
			tests, e := store.Tests{}.ListAll(ctx, sc, 0)
			if e != nil {
				return e
			}
			for _, t := range tests {
				existing = append(existing, t.Target)
			}
			incs, e := store.Incidents{}.List(ctx, sc)
			if e != nil {
				return e
			}
			counts := map[string]int{}
			for _, inc := range incs {
				target := inc.Target
				if target == "" {
					target = inc.Prefix
				}
				if target != "" {
					counts[target]++
				}
			}
			for target, n := range counts {
				obs = append(obs, author.Observation{Target: target, Kind: "incident", Count: n})
			}
			return nil
		}); err != nil {
			return err
		}
	}
	// Incident targets are already correlated signals (low noise), so a single
	// occurrence is worth proposing.
	proposals := author.Discover(obs, existing, author.DiscoverOptions{MinCount: 1})
	writeJSON(w, http.StatusOK, map[string]any{"proposals": proposals})
	return nil
}
